src/main/java/io/axoniq/axonserver/connector/query/impl/QueryChannelImpl.java

```diff
@@ -601,6 +622,37 @@ public void sendAck() {
         );
     }
 
+    private static class QueryInProgress {
+
+        private final CompletableFuture<?> closeHandler;
+        private final QueryHandler.FlowControl flowControl;
+
+        public static QueryInProgress noop() {
+            return new QueryInProgress(() -> { }, NoopFlowControl.INSTANCE);
+        }
+
+        public QueryInProgress(Runnable closeHandler, QueryHandler.FlowControl flowControl) {
+            this.closeHandler = new CompletableFuture<>();
+            this.flowControl = flowControl;
+            this.closeHandler.whenComplete((r, e) -> {
+                flowControl.complete();
+                closeHandler.run();
+            });
+        }
+
+        public CompletableFuture<?> completeAsync() {
+            return closeHandler;
```

Comment:
This method doesn't actually complete anything. Instead, it returns the handle that is triggered with it does.
Suggestion: rename to "whenComplete". This would give code like:
`queryInProgress.whenComplete().thenRun(....)`